Bugfix/77 wpcom vip add role caps fix

ovidiul opened the pull request:
It fixes the issue mentioned in #77  by renaming the activation hook to admin_init and adding a version check to only init the roles once when the version changes as per https://docs.wpvip.com/how-tos/customize-user-roles/ .

PR #93 might need to be merged before this.

GaryJones commented on .gitignore:
```diff
@@ -4,3 +4,4 @@
 /phpcs.xml
 /phpunit.xml
 /.phpunit.result.cache
+node_modules
```
Unrelated change.

ovidiul replied:
removed

GaryJones commented on README.md:
```diff
@@ -21,6 +21,30 @@ Please contact us before using this plugin.
 - PHP 5.6+
 - WordPress 4.5+
 
+## Developement - Getting Started
```
Unrelated change.

GaryJones commented on composer.json:
```diff
@@ -1,5 +1,4 @@
-{
-  "name": "automattic/wpcom-legacy-redirector",
+{                                                                                                                                                                                                                              "name": "automattic/wpcom-legacy-redirector",
```
Unrelated change.

GaryJones commented on composer.json:
```diff
@@ -41,3 +40,4 @@
     ]
   }
 }
+
```
Don't include `composer.lock`.

GaryJones commented on includes/class-wpcom-legacy-redirector-cli.php:
```diff
@@ -351,13 +351,15 @@ public function export_to_csv( $args, $assoc_args ) {
-				'post_status'      => 'any',
-				'suppress_filters' => 'false',
-			) );
+			$posts = get_posts(
```
Unrelated change.

GaryJones commented on includes/class-wpcom-legacy-redirector-ui.php:
```diff
@@ -125,10 +125,25 @@ public function custom_vip_legacy_redirect_column( $column, $post_id ) {
+	 *
+	 * @return string
+	 */
+	private function get_home_domain_without_path() {
```
Unrelated change - fork from `develop` and apply the changes to that new branch.

GaryJones commented on includes/class-wpcom-legacy-redirector.php:
```diff
@@ -6,13 +6,13 @@
 	const POST_TYPE   = 'vip-legacy-redirect';
 	const CACHE_GROUP = 'vip-legacy-redirect-2';
-
+	const VERSION     = '1.4.0';
```
We've not tagged `1.4.0` yet, so this can't be correct.

ovidiul replied:
fixed to latest known stable 1.3.0 for now

GaryJones commented on includes/class-wpcom-legacy-redirector.php:
```diff
@@ -64,6 +64,14 @@ static function init() {
+		$plugin_version_key = self::POST_TYPE . '_version';
+
+		// Check if this has been run already.
+		if ( self::VERSION <= get_option( $plugin_version_key ) ) {
```
The use of the plugin version here means that it will redundantly run on each new plugin version. Instead, we can use a different constant that has an integer just related to the version of these roles and caps changes.

ovidiul replied:
I've added a new CAPABILITIES_VER constant for it to fix this with value set to 1.0

GaryJones commented on includes/class-wpcom-legacy-redirector.php:
```diff
@@ -328,7 +339,14 @@ static function get_redirect_uri( $url ) {
 			} elseif ( ! empty( $redirect_post->post_excerpt ) ) {
 				// Add preserved params to the destination URL.
-				return add_query_arg( $preservable_params, esc_url_raw( $redirect_post->post_excerpt ) );
+				// We need to add here the home_url() if the target starts with.
```
Unrelated change.

GaryJones commented on includes/class-wpcom-legacy-redirector.php:
```diff
@@ -365,6 +383,46 @@ private static function get_url_hash( $url ) {
+	 *               requested: null if the component doesn't exist in the given URL; a
+	 *               string (or in the case of PHP_URL_PORT, integer) when it does. 
+	 */
+	public static function mb_parse_url( $url, $component = -1 ) {
```
Unrelated change.

GaryJones commented on package.json:
```diff
@@ -0,0 +1,26 @@
+{
```
Unrelated change.

GaryJones commented on tests/integration/RedirectsTest.php:
```diff
@@ -6,6 +6,23 @@ class RedirectsTest extends TestCase {
 	public function get_redirect_data() {
 		return array(
+			'redirect_relative_path'  => array(
+				'/non-existing-page',
```
Unrelated change.

GaryJones commented on wpcom-legacy-redirector.php:
```diff
@@ -25,7 +25,7 @@
 
 if ( defined( 'WP_CLI' ) && WP_CLI ) {
-	require __DIR__ . '/includes/class-wpcom-legacy-redirector-cli.php';
+	include __DIR__ . '/includes/class-wpcom-legacy-redirector-cli.php';
```
Unrelated change.

GaryJones reviewed (changes requested):
There are too many unrelated changes in this PR. The actual fix needs about 4 lines of code, and a unit or integration test.

I have a local branch that extracts the logic for handling the capability into a new class, so I'll get that pushed and then we can add the logic and new constant there.

ovidiul committed:
Capability: Reduce registration instances

- Change activation hook from `init` to `admin_init`.
- Adds version check to only init the capability registration once, as per https://docs.wpvip.com/how-tos/customize-user-roles/.
- Adds a convenience method to unregister the capability from specific roles as well.
- Includes unit and integration tests for the whole capability registration process

Fixes #77.

ovidiul commented:
Hi Gary, I've rebased my PR and added the requested changes.

Let me know if you see further adjustments needed.

GaryJones commented on includes/class-wpcom-legacy-redirector.php:
```diff
@@ -13,27 +13,45 @@ class WPCOM_Legacy_Redirector {
+	const PLUGIN_VERSION = '1.3.0';
+
+	// Used to flip the version of the available roles capabilities.
+	const CAPABILITIES_VER = '1.0';
```
You're doing a numerical comparison, but have this set as a string? Use an int instead.

```suggestion
	const CAPABILITIES_VER = 1;
```

ovidiul replied:
done

GaryJones commented on includes/class-wpcom-legacy-redirector.php:
```diff
@@ -13,27 +13,45 @@ class WPCOM_Legacy_Redirector {
+	static function admin_init() {
+		$capabilities_version_key = self::POST_TYPE . '_capabilities_version';
+
+		if ( self::CAPABILITIES_VER <= get_option( $capabilities_version_key ) ) {
```
I think this logic is wrong. Currently it says if the constant is less then or equal to the saved version number (which it should be for much of the time), then register the new capability.

I _think_ this should be:

```suggestion
		if ( self::CAPABILITIES_VER > get_option( $capabilities_version_key ) ) {
```

The integration tests below check that the capability is added to the roles, but it doesn't check how many times this `Capability->register()` command is run, which is part of what we're  trying to fix here.

ovidiul replied:
found a way to test this, I've moved related code to the capability class and added a return statement in that function, this would give us an easy way to test the register call after the initial init

GaryJones commented on tests/integration/CapabilityTest.php:
```diff
@@ -0,0 +1,76 @@
+	protected static $initialized = false;
+
+	public function setUp() {
+		if ( ! self::$initialized ) {
+			$capability = new Capability();
+			$capability->register();
+
+			self::$initialized = true;
+		}
```
To test the logic in `includes/class-wpcom-legacy-redirector.php` (about whether the Capability method is called or not), then you may want to have this moved to each method instead of in this setUp.

If anything is left in the setUp, then you may need a tearDown as well.

ovidiul replied:
I've added the logic, but replaced the setUp with a beforeClass, we could skip this but I want to make sure we have no capabilities setup before the tests.

I use tearDown to check the register init and making sure it's not being setup again after the initial init.

GaryJones commented on tests/integration/CapabilityTest.php:
```diff
@@ -0,0 +1,76 @@
+	 */
+	public function test_admin_capability() {
+
+		$user_id = $this->factory->user->create( array( 'role' => 'administrator' ) );
+		$user    = wp_set_current_user( $user_id );
```
Instead of testing a _new_ administrator, can you also check to see if an existing administrator account has got the cap? (Perhaps another reason to move the `Capability->register()` to these test methods so you can do a before and after check.

ovidiul replied:
I modified the tests accordingly, added also an unregister method with the relevant test in the Capability section.

GaryJones reviewed (changes requested):
I'd like to see more variety on the integration tests to confirm that the caps are not present on the various roles, then are added and are present, and that they are not re-added unnecessarily on subsequent requests as well.

GaryJones commented on tests/integration/CapabilityTest.php:
```diff
@@ -72,12 +86,37 @@ public function test_subscriber_capability() {
+
+		$this->assertFalse( array_key_exists( Capability::MANAGE_REDIRECTS_CAPABILITY, get_role( 'administrator' )->capabilities ) );
+
+		$capability->register();
```
Is this final register really needed?

ovidiul replied:
Yes, it was needed for the tearDown, however I've refactored that now to not require it.

ovidiul replied:
not anymore, it was needed for the tearDown method

GaryJones commented on tests/integration/CapabilityTest.php:
```diff
@@ -0,0 +1,131 @@
+	 *
+	 * @beforeClass
+	 */
+	public static function beforeClassSetup() {
+		$capability = new Capability();
+		$capability->unregister();
+	}
```
Is this needed? (It may be!)

As part of testing, we register the plugin as being activated, but that shouldn't trigger the `admin_init` hook which may do the cap registering.

If other test classes run first, I don't think it should be left in a caps-registered state.

If this method is scrapped, do the tests pass? If so, then it either it was redundant, or there's another thing to be tested.

ovidiul replied:
It seems it is, what happens is that after each test, the wp_options tests table retains it's roles option to the mentioned state, so I think it's best after each test to remove any set capabilities

GaryJones commented on tests/integration/CapabilityTest.php:
```diff
@@ -0,0 +1,131 @@
+	}
+
+	public function test_admin_capabilities_before_register() {
+		$user_id = $this->factory->user->create( array( 'role' => 'administrator' ) );
+		$user    = wp_set_current_user( $user_id );
```
As as separate test, can you grab user ID 1 and check their caps before and after registering?

So rename this test to `test_new_admin_capabilities_before_register` and create a `test_existiing_admin_capabilities_before_register` test method?

GaryJones replied:
We'll also want to see tests for when `wpcom_vip_add_role_caps()` exists and when it doesn't since right now I suspect only one of those branches is being tested.

ovidiul replied:
I've added the ID=1 user test, but for multisite, this won't work because of in WP_User class, on multisite the has_cap [returns true](https://github.com/WordPress/WordPress/blob/master/wp-includes/class-wp-user.php#L754-L760) for admin users

Not sure about wpcom_vip_add_role_caps , we might need to mock it, but i think we should do that in a separate PR?

GaryJones commented on tests/integration/CapabilityTest.php:
```diff
@@ -0,0 +1,131 @@
+		$user_id = $this->factory->user->create( array( 'role' => 'administrator' ) );
+		$user    = wp_set_current_user( $user_id );
+
+		$this->assertFalse( boolval( $user->has_cap( Capability::MANAGE_REDIRECTS_CAPABILITY ) ) );
```
`->has_cap()` only returns a boolean `true` or `false`, so `boolval()` is redundant here.

ovidiul replied:
This is removed now.

GaryJones commented on tests/integration/CapabilityTest.php:
```diff
@@ -0,0 +1,131 @@
+		$user_id = $this->factory->user->create( array( 'role' => 'administrator' ) );
+		$user    = wp_set_current_user( $user_id );
+
+		$this->assertTrue( $user->has_cap( Capability::MANAGE_REDIRECTS_CAPABILITY ) );
```
Let's wrap these assertions into custom assertion methods at the end of this test class. Something like:

```php
protected function assertUserHasRedirectCapability( $user ) {
	return $this->assertTrue( $user->has_cap( Capability::MANAGE_REDIRECTS_CAPABILITY ) );
}

protected function assertUserNotHasRedirectCapability( $user ) {
	return $this->assertFalse( $user->has_cap( Capability::MANAGE_REDIRECTS_CAPABILITY ) );
}

protected function assertRoleHasRedirectsCapability( $role ) {
	$user_id = $this->factory->user->create( array( 'role' => $role ) );
	$user    = wp_set_current_user( $user_id );

	return $this->assertUserHasRedirectCapability( $user );
}

protected function assertRoleNotHasRedirectsCapability( $role ) {
	$user_id = $this->factory->user->create( array( 'role' => $role ) );
	$user    = wp_set_current_user( $user_id );

	return $this->assertUserNotHasRedirectCapability( $user );
}
```

Then in the test methods, you can do:

```php
public function test_new_admin_capability() {
	$this->assertRoleNotHasRedirectsCapability( 'administrator' );

	$capability        = new Capability();
	$register_response = $capability->register();

	$this->assertRoleHasRedirectsCapability( 'administrator' );	
}
```

This could then be extracted to use a dataProvider since then the only variable is the role - wrap the final assertion in a condition that either checks the role has, or role has not got he cap, based on the expectation in the dataProvider.

Alternatively, just bundle the different role checks together:

```php
public function test_new_admin_capability() {
	$this->assertRoleNotHasRedirectsCapability( 'administrator' );
	$this->assertRoleNotHasRedirectsCapability( 'editor' );
	$this->assertRoleNotHasRedirectsCapability( 'subscriber' );

	$capability        = new Capability();
	$register_response = $capability->register();

	$this->assertRoleHasRedirectsCapability( 'administrator' );
	$this->assertRoleHasRedirectsCapability( 'editor' );
	$this->assertRoleNotHasRedirectsCapability( 'subscriber' ); // Should be no change.
}
```

ovidiul replied:
New methods added and the necessary code refactored

GaryJones commented on includes/class-capability.php:
```diff
@@ -5,10 +5,21 @@
 	const MANAGE_REDIRECTS_CAPABILITY = 'manage_redirects';
 
+	// Used to flip the version of the available roles capabilities.
+	const CAPABILITIES_VER = '1';
```
This is still a string? It's used in line 19 with a `<=` operator.

ovidiul replied:
Fixed now

GaryJones commented on includes/class-capability.php:
```diff
@@ -19,5 +30,34 @@ public function register() {
+	 */
+	public function unregister() {
+
+		$capabilities_version_key = self::MANAGE_REDIRECTS_CAPABILITY . '_capability_version';
```
This is the second time this line is used. Let's extract it into its own method for consistency.

ovidiul replied:
done

GaryJones commented on includes/class-wpcom-legacy-redirector.php:
```diff
@@ -13,27 +13,36 @@ class WPCOM_Legacy_Redirector {
+	 */
+	static function admin_init() {
+		$capability = new Capability();
+			$capability->register();
```
Incorrect number of indentation tabs.

ovidiul replied:
fixed

GaryJones commented on tests/integration/CapabilityTest.php:
```diff
@@ -0,0 +1,130 @@
+		$this->assertTrue( $capability->register() );
+	}
+
+	/**
+	 * Test if the capabilities register function fires again after init
+	 *
+	 * @return void
+	 */
+	public function tearDown() {
+
+		$capability = new Capability();
+		$capability->unregister();
+	}
```
Move this fixture to the top, and check the DocBlock description.

ovidiul replied:
fixed

GaryJones reviewed (changes requested):
Looking better, but I'm still not seeing a test that checks that update_option() was only called once when `$capability->register()` called twice. Have a look at https://brain-wp.github.io/BrainMonkey/docs/functions-expect.html which allows you to test for exactly this. Set the `Function\expects`, ensure the saved option is empty, call register twice, and see if the test is green (Might need to put that into a unit test, rather than the integration test)

ovidiul commented:
> Looking better, but I'm still not seeing a test that checks that update_option() was only called once when `$capability->register()` called twice. Have a look at [brain-wp.github.io/BrainMonkey/docs/functions-expect.html](https://brain-wp.github.io/BrainMonkey/docs/functions-expect.html) which allows you to test for exactly this. Set the `Function\expects`, ensure the saved option is empty, call register twice, and see if the test is green (Might need to put that into a unit test, rather than the integration test)

I've added a unit test in tests/unit/CapabilityTest.php

We do have a related test in `tests/integration/CapabilityTest.php` , method `test_register` which checks the return of the register method, I guess another one doesn't hurt.

GaryJones commented on tests/unit/CapabilityTest.php:
```diff
@@ -0,0 +1,65 @@
+ */
+class CapabilityUnitTest extends TestCase {
+
+	/**
+	 * Setup before each test
+	 *
+	 * @return void
+	 */
+	protected function setUp() {
+		parent::setUp();
+		Monkey\setUp();
+	}
+
+	/**
+	 * TEar down after each test
+	 *
+	 * @return void
+	 */
+	protected function tearDown() {
+		Monkey\tearDown();
+		parent::tearDown();
+	}
```
Because this class extends `Yoast\WPTestUtils\BrainMonkey\TestCase` then I believe these Monkey setup and teardown is already handled - see https://github.com/Yoast/wp-test-utils/blob/develop/src/BrainMonkey/TestCase.php

ovidiul replied:
even better, i've removed that now.

GaryJones commented on composer.json:
```diff
@@ -16,7 +16,8 @@
     "phpunit/phpunit": "^5 || ^6 || ^7",
-    "yoast/wp-test-utils": "^0.2"
+    "yoast/wp-test-utils": "^0.2",
+    "brain/monkey": "^2.6"
```
Brain Monkey is already included with wp-test-utils.

GaryJones commented:
I've squashed all of the commits, and rebased on to `develop`, so we'll have a clean history.

GaryJones reviewed (approved):
LGTM!
